### ISSUE TITLE
VirtualBox cannot run alongside WSL 2

### ISSUE THREAD

**Issue by LjMario007** [myspaghetti/macos-guest-virtualbox#157]
I'm running WSL 2 on the latest Windows Insider Preview build.

When I ran the script, it worked correctly up until booting the VM, at which point VirtualBox threw up an error (`VERR_NEM_VM_CREATE_FAILED`). Attempting to fix this issue led me to [this StackOverflow post](https://stackoverflow.com/questions/58031941/how-to-get-oracle-vm-virtualbox-6-0-and-wsl-working-at-the-same-time), the solution of which said to run `bcdedit /set hypervisorlaunchtype off` in an elevated `cmd` window to fix the problem, and reboot.

This worked to get VirtualBox running, but when I went to run the script again, WSL wouldn't run, saying `Please enable the Virtual Machine Platform Windows feature and ensure virtualization is enabled in the BIOS.` Investigating further, I found that WSL 2 is incompatible with VirtualBox as per @earonesty in [this issue in the WSL repository](https://github.com/MicrosoftDocs/WSL/issues/404#issuecomment-549381790):

> Also, note: upgrading to WSL2 this way will break VirtualBox and probably other virtualization engines. WSL2 is no longer a lightweight compatibility system... instead it's a full VM, with all the mount issues, etc of VM's.

[Later in the discussion](https://github.com/MicrosoftDocs/WSL/issues/404#issuecomment-568616190), @hrai said to run `bcdedit /set hypervisorlaunchtype auto start` in an elevated `cmd` window to fix WSL not working, but this then makes VirtualBox stop working and throw up the `VERR_NEM_VM_CREATE_FAILED` error above.

So basically this means that when WSL 2 is finally released out of the Windows Insider Program as part of a future Windows update, the WSL portion of this Bash script will no longer work because WSL 2 (permanently) breaks compatibility with VirtualBox. I can't say as to whether the Cygwin portion will continue to work (it probably will, seeing as it's separate from WSL), but I wanted to give you a heads up that WSL will no longer work with this script in the future.

**Comment by myspaghetti** [comment]
Didn't the script show you the following warning?

>The script is not tested with WSL2, which uses Hyper-V.
>VirtualBox Hyper-V support is experimental.

If you don't mind, please run on the WSL2 terminal `cat /proc/sys/kernel/osrelease` and post the output here.

**Comment by LjMario007** [comment]
I have tried the script with Hyper-V both enabled and disabled in VirtualBox.
`cat /proc/sys/kernel/osrelease` says the following:
`4.19.84-microsoft-standard`

**Comment by myspaghetti** [comment]
The script should have warned you about WSL2, unless you have Cygwin installed in addition to WSL2, in which case I'm not about to tackle that corner-case.

Either way I made the warning more explicit in [version 0.86.1](https://github.com/myspaghetti/macos-guest-virtualbox/commit/832c161a17af5d0cc7a33942d23b42eefa6ca14d)

**Comment by LjMario007** [comment]
Okay!

One question: when WSL 1 is phased out and gradually replaced with WSL 2, will this script become Cygwin-only seeing as WSL 2 and VirtualBox cannot run at the same time?

**Comment by myspaghetti** [comment]
Microsoft said WSL1 will not be phased out in the near future. Hopefully VirtualBox will properly support Hyper-V by the time WSL1 is phased out, and/or Microsoft fixes their Hyper-V implementation so it doesn't block other virtualization software.
